mllib/src/main/scala/org/apache/spark/ml/feature/MinMaxScaler.scala

```diff
@@ -0,0 +1,171 @@
+    transformSchema(dataset.schema, logging = true)
+    val input = dataset.select($(inputCol)).map { case Row(v: Vector) => v }
+    val summary = input.treeAggregate(new MultivariateOnlineSummarizer)(
+      (aggregator, data) => aggregator.add(data),
+      (aggregator1, aggregator2) => aggregator1.merge(aggregator2))
+
+    copyValues(new MinMaxScalerModel(uid, summary.min, summary.max).setParent(this))
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    val inputType = schema($(inputCol)).dataType
+    require(inputType.isInstanceOf[VectorUDT],
+      s"Input column ${$(inputCol)} must be a vector column")
+    require(!schema.fieldNames.contains($(outputCol)),
+      s"Output column ${$(outputCol)} already exists.")
+    val outputFields = schema.fields :+ StructField($(outputCol), new VectorUDT, false)
+    StructType(outputFields)
+  }
+
+  override def validateParams(): Unit = {
+    require($(min) <= $(max))
+  }
+
+}
+
+/**
+ * :: AlphaComponent ::
+ * Model fitted by [[MinMaxScaler]].
+ */
+@AlphaComponent
+class MinMaxScalerModel private[ml] (
+    override val uid: String,
+    val originMin: Vector,
+    val originMax: Vector)
+  extends Model[MinMaxScalerModel] with MinMaxScalerParams {
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    transformSchema(dataset.schema, logging = true)
+
+    val scale = udf { (vector: Vector) => {
+      val scale = $(max) - $(min)
+      val result = vector match {
+        case DenseVector(vs) =>
+          val values = vs.clone()
+          val size = values.size
+          var i = 0
+          while(i < size) {
+            val originalRange = originMax(i) - originMin(i)
+            val raw = if(originalRange != 0) (values(i) - originMin(i)) / originalRange else 0.5
+            values(i) =  raw * scale + $(min)
+            i += 1
+          }
+          Vectors.dense(values)
+        case SparseVector(size, indices, vs) =>
+          // For sparse vector, the `index` array inside sparse vector object will not be changed,
+          // so we can re-use it to save memory.
+          val values = vs.clone()
+          val nnz = values.size
+          var i = 0
+          while (i < nnz) {
+            val index = indices(i)
+            val originRange = originMax(index) - originMin(index)
+            val raw = if(originRange != 0) (values(i) - originMin(index)) / originRange else 0.5
+            values(i) = raw  * scale + $(min)
+            i += 1
+          }
+          Vectors.sparse(size, indices, values)
+        case v => throw new IllegalArgumentException("Do not support vector type " + v.getClass)
+      }
+      result
+    }}
+    dataset.withColumn($(outputCol), scale(col($(inputCol))))
```

Comment:
We should really update the ML metadata.  Can you please look at other feature transformer examples for that?  That could come in a follow-up PR, but if so, we should go ahead and create the JIRA + add a warning in the Scala doc.